src/Stratis.Bitcoin.Features.Wallet.Tests/WalletSyncManagerTest.cs

```diff
@@ -98,21 +102,23 @@ public void ProcessBlock_NewBlock_PreviousHashSameAsWalletTip_PassesBlockToManag
             this.chain = result.Chain;
             List<Block> blocks = result.Blocks;
             var walletSyncManager = new WalletSyncManagerOverride(this.LoggerFactory.Object, this.walletManager.Object, this.chain, KnownNetworks.StratisMain,
-                this.blockStoreCache.Object, this.storeSettings, this.nodeLifetime.Object);
+                this.blockStore.Object, this.storeSettings, this.nodeLifetime.Object);
             walletSyncManager.SetWalletTip(this.chain.GetBlock(3));
 
             Block blockToProcess = blocks[3];
             walletSyncManager.ProcessBlock(blockToProcess); //4th block in the list has same prevhash as which is loaded
 
-            uint256 expectedBlockHash = this.chain.GetBlock(4).Header.GetHash();
-            Assert.Equal(expectedBlockHash, walletSyncManager.WalletTip.Header.GetHash());
+            uint256 expectedBlockHash = this.ExpectedBlockHash(walletSyncManager, 4);
```

Comment:
Is this correct?

Reply:
Yes, before adding this line the test would call `ProcessBlock()`, enqueue the block, then drop out the method - which is correct.  

However, this would invoke the callback `WalletSyncManager.OnProcessBlockAsync()` to determine if the block needs to be processed - and at the same time the test would start to Assert or Verify and fail.

To get around this the Assert on the expected block is wrapped in a `WaitLoop()` - which is now called `AssertTipBlockHash()`.  This does not return until the Func predicate returns true, or false after 30 seconds.

This is a better approach then using `Thread.Sleep()`.